Internal documentation for iptables rules

robmry opened the pull request:
**- What I did**

Generate documentation describing how the `iptables` rules fit together, and where they come from in the code.

The output is at https://github.com/robmry/moby/blob/iptables_doc/integration/network/bridge/iptablesdoc/index.md

A lot more description could be added, particularly for the `nat` table, as well as additional network configurations. Initially, I've concentrated on what's needed to document upcoming changes.

_This documentation is likely to get less useful as progress is made on refactoring and the code gets easier to understand and describe in comments. But, in the meantime, it'll be useful to check for unexpected changes as we refactor._

**- How I did it**

An integration test:
  - runs daemons in their own network namespaces, which then
  - create networks in various configurations, and
  - dump iptables
  - combines those dumps with `text/template` files containing GitHub-flavoured markdown, and finally
  - compares the result with checked-in golden reference files.

So, if the iptables rules change, the integration test will fail. The template text may then need updates, and the test can be run with `TESTFLAGS="-update"` to update the generated files.

**- How to verify it**

Run the test, read the generated docs.

**- Description for the changelog**
```markdown changelog
n/a
```

robmry committed:
Fix iptables rule descriptions

Signed-off-by: Rob Murray <rob.murray@docker.com>

AkihiroSuda commented on cmd/dockerd/daemon.go:
```diff
@@ -354,13 +354,19 @@ func (cli *daemonCLI) start(ctx context.Context) (err error) {
 		log.G(ctx).WithError(err).Error("Failed to close buildkit")
 	}
 
+	log.G(ctx).Info("closed buildkit")
```
Accidentally added an irrelevant commit?

robmry replied:
Not accidental - this PR's still draft, and that's part of temporary commit [TEMP: dump daemon logs after d.Stop (debugging error-exit)](https://github.com/moby/moby/pull/48593/commits/205569bee1a2ea0642ba76acc445d824954e6482) ...

For some reason, the daemon's not shutting down cleanly in CI (it's fine in a dev env). I think it may be something to do with OTEL tracing, because with that extra debug the daemon log ends with ...

```
        time="2024-10-08T09:26:07.239164944Z" level=debug msg="Clean shutdown succeeded"
        time="2024-10-08T09:26:07.239249993Z" level=info msg="closed buildkit"
        time="2024-10-08T09:26:07.239258238Z" level=info msg=cancelled
        time="2024-10-08T09:26:07.239262526Z" level=info msg="errAPI !ok"
        time="2024-10-08T09:26:27.238043935Z" level=info msg="Processing signal 'interrupt'"
        time="2024-10-08T09:26:28.237981737Z" level=info msg="Processing signal 'interrupt'"
        time="2024-10-08T09:26:29.238348730Z" level=info msg="Processing signal 'interrupt'"
        time="2024-10-08T09:26:29.238384185Z" level=info msg="Forcing docker daemon shutdown without cleanup; 3 interrupts received"
```

So, it looks like it didn't come back from [tp.Shutdown](https://github.com/moby/moby/blob/205569bee1a2ea0642ba76acc445d824954e6482/cmd/dockerd/daemon.go#L370).), but I'm still investigating ...

robmry replied:
There's no routing set up in the netns where the daemon's running during this test, so the OTEL endpoint isn't reachable and that misconfiguration causes the daemon to get stuck during exit ... not sure it should, but that's a separate discussion.

For now, I've disabled OTEL for this test, and got rid of that temporary commit.

AkihiroSuda commented on libnetwork/drivers/bridge/port_mapping_linux.go:
```diff
@@ -802,7 +802,7 @@ func setPerPortForwarding(b portBinding, ipv iptables.IPVersion, bridgeName stri
 	}
 	rule := iptRule{ipv: ipv, table: iptables.Filter, chain: DockerChain, args: args}
-	if err := appendOrDelChainRule(rule, "MASQUERADE", enable); err != nil {
+	if err := appendOrDelChainRule(rule, "OPEN PORT", enable); err != nil {
```
Why change?

thaJeztah replied:
These functions are horrible; from the "before" code, I really assumed `MASQUERADE` was some special value required for the code to  work (wondered "can we just use _two words_ here?". But ... all it is is a custom message to include in an error that's produced. It's just an abstraction to not have to write an error message here 🤦 
 https://github.com/moby/moby/blob/8cc4029162401663ae10be601ffb69d9cf2dfb46/libnetwork/drivers/bridge/setup_ip_tables_linux.go#L336

thaJeztah replied:
I think I actually may have a branch somewhere that changed these. But can't find them 😂

robmry committed:
Generate iptables documentation

In an integration test - run a daemon, capture iptables, and feed them
to a markdown text/template describing them.

Prep for repeating that, for different network configurations.

Fail the test if the generated markdown differs from a "golden" version.

(So, at-least the golden markdown will need to be updated if the
iptables rules are deliberately changed - hopefully the corresponding
description in the template will also be updated.)

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Add iptables doc for a user-defined network

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Add iptables doc for a network with userland-proxy disabled

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Add iptables doc for a network with icc=false

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Add iptables doc for a --internal network

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Add iptables doc for a gateway-mode=routed network

Signed-off-by: Rob Murray <rob.murray@docker.com>

unclejack commented:
@robmry: There are a few scenarios we may want to cover with these tests:
- default bridge network with user land proxy disabled
- forwarded port ranges (with the same port ranges on the host and the container side, with different ports, e.g.: 8000-8004:8000-8004, 8000-8004:10000-10004)
- various combinations of these and other configurations already covered

robmry commented:
> @robmry: There are a few scenarios we may want to cover with these tests:

Thank you yes, definitely. Also '--link' for the default bridge, disabling masquerading, and I'm sure a whole lot of other special cases.

This initial set is what I needed to document the changes in the current series of PRs. Hopefully it'll make those reviews a bit easier.

Once we've got the special cases covered here, I think we'll be able to get rid of quite a few unit tests that are hard to maintain because the have code to check for specific things (but not generally the order of rules etc).

tianon commented:
I was gushing about this privately (due to seeing the results of it on one of @robmry's stacked PRs), but I think this is an _amazing_ idea -- reviewing the impact of https://github.com/moby/moby/commit/b53bdcecfbfd5fc749a93961a046ae0b9af8d2ba is so trivial with this change :face_holding_back_tears:

akerouanton reviewed (commented):
Really great work! Thanks!

I'm wondering if it'd be better to include a diff of iptables rules for `usernet-portmap-*.md` against the baseline (ie. `usernet-portmap.md`) 🤔

robmry commented on integration/network/bridge/iptablesdoc/templates/usernet-internal.md:
```diff
@@ -0,0 +1,43 @@
+  - Rule 1 drops any packet routed to the network that does not have a source address in the network's subnet.
+  - Rule 2 drops any packet routed out of the network that does not have a dest address in the network's subnet.
+  - There is no jump to DOCKER-ISOLATION-STAGE-2.
+- DOCKER-ISOLATION-STAGE-1 is unused.
```
Fixed.

robmry commented on integration/network/bridge/iptablesdoc/templates/usernet-portmap-noproxy.md:
```diff
@@ -0,0 +1,47 @@
+    [ProgramChain][1].
+  - The "SKIP DNAT" RETURN rule for packets routed to the bridge is omitted from
+    the DOCKER chain [setupIPTablesInternal][2].
+  - The MASQUERADE rule for packets from LOCAL source address is not included in
```
Fixed.

robmry commented on integration/network/bridge/iptablesdoc/templates/usernet-portmap-noicc.md:
```diff
@@ -0,0 +1,39 @@
+## Container on a user-defined network with inter-container communication disabled, with a published port
+
+Equivalent to:
```
I've enabled the proxy in all but the noproxy test.

robmry commented on integration/network/bridge/iptablesdoc/templates/usernet-internal.md:
```diff
@@ -0,0 +1,43 @@
+## Container on a user-defined --internal network
+
+Equivalent to:
```
The proxy's enabled now.

robmry commented on integration/network/bridge/iptablesdoc/iptablesdoc_linux_test.go:
```diff
@@ -122,6 +122,19 @@ var index = []section{
 	},
+	{
+		name: "usernet-portmap-routed.md",
+		networks: []bridgeNetwork{{
```
I've enabled the proxy in all but the noproxy test.

robmry commented:
> I'm wondering if it'd be better to include a diff of iptables rules for `usernet-portmap-*.md` against the baseline (ie. `usernet-portmap.md`) 🤔

Yes, possibly ... at the moment, there's a test per section of the doc that builds the network and dumps the iptables. So, generating a diff from a baseline setup would introduce a dependency between tests, which could get messy. Alternatively, each test could run the baseline config as well as its own.

But, now the userland-proxy is running in most of the configs, it should be a bit easier to eyeball the changes. Perhaps it's something to add later if some very subtle differences need to be explained.

akerouanton reviewed (approved):
LGTM

laurazard commented on internal/testutils/networking/l3_segment_linux.go:
```diff
@@ -114,14 +116,14 @@ func newHost(t *testing.T, nsName, ifname string) Host {
 }
 
 // Run executes the provided command in the host's network namespace.
```
```suggestion
// Run executes the provided command in the host's network namespace and returns it's output.
```

robmry replied:
Done.

laurazard commented on integration/network/bridge/iptablesdoc/templates/usernet-portmap-routed.md:
```diff
@@ -0,0 +1,53 @@
+
+Differences from [nat mode][1]:
+
+  - In the POSTROUTING chain:
+    - No MASQUERADE rule for traffic from the bridge network to elsewhere. [setupIPTablesInternal][2]
+    - No MASQUERADE rule for traffic from the bridge network to itself on published port 80 (port
+      mapping is skipped). [attemptBindHostPorts][3]
+  - In the DOCKER chain:
+    - No early return ("skip DNAT") for traffic from the bridge network. [setupIPTablesInternal][4]
+    - No DNAT rule for the published port (port mapping is skipped). [attemptBindHostPorts][3]
+
+_And, the userland proxy won't be started for mapped ports._
```
I wonder if we can generate this/include this from the test instead of just having it directly in the template. Benefit would be to have one less place to have to update when making changes.

Not a blocker for sure, just a thought for future improvements.

robmry replied:
Yes, there's definitely scope for putting more of the words into the code itself.

In particular, the links have a chicken-and-egg problem - no way to get a permalink to the code until it's pushed, so there's no way to link to changed code. In later PRs, I linked to code in my clone or removed links and just named functions ... when this series of iptables changes are merged I'll make a change to update links.

Once we've refactored, ready to add native nftables support, the code should be less scattered and more stable. Then this should all get a bit easier.

laurazard reviewed (approved):
This looks really good, thanks!!
